src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumClient.kt

```diff
@@ -12,237 +12,275 @@ import kotlinx.coroutines.flow.*
-                        closeSocket(null)
-                    }
                 }
             }
-        }
 
-        suspend fun listen() {
-            try {
+            val listenJob = launch(CoroutineName("listen")) {
+                val flow = linesFlow(socket).map { json.decodeFromString(ElectrumResponseDeserializer, it) }
                 flow.collect { response -> mailbox.send(Action.ProcessServerResponse(response)) }
-                closeSocket(null)
-            } catch (ex: TcpSocket.IOException) {
-                logger.warning { "TCP receive: ${ex.message}" }
-                closeSocket(ex)
             }
+
+            // Suspend until the coroutine is cancelled or the socket is closed.
+            listenJob.join()
         }
 
-        launch { ping() }
-        launch { respond() }
+        return ListenJob(job, socket)
+    }
 
-        listen() // This suspends until the coroutine is cancelled or the socket is closed
+    /** This can be used to set longer timeouts if we detect a slow connection (e.g. Tor). */
+    fun setRpcTimeout(timeout: Duration) {
+        rpcTimeout = timeout
     }
 
-    override suspend fun send(request: ElectrumRequest, replyTo: CompletableDeferred<ElectrumResponse>) {
+    /**
+     * We may never get a response to our requests, because the electrum protocol is fully asynchronous (we write a
+     * request on the TCP socket, and at some point in the future the server may send us a response).
+     *
+     * This can for example happen if:
+     *  - the server drops our request for whatever reason
+     *  - we get disconnected after sending the request and reconnect to a different server
+     *
+     * We can mitigate that by retrying after a delay and disconnecting from servers that don't seem to be responsive.
+     *
+     * @param replyTo callers should use the same replyTo for retries: this guarantees that if a previous attempt
+     * succeeds after its timeout, we will return the result and ignore the retries.
+     */
+    private suspend inline fun rpcCallWithTimeout(replyTo: CompletableDeferred<ElectrumResponse>, request: ElectrumRequest): ElectrumResponse? {
+        // Sending to the mailbox suspends until the request is picked up and written to the TCP socket.
+        // We explicitly keep this out of the timeout, which lets us suspend until we're connected to an electrum server.
         mailbox.send(Action.SendToServer(Pair(request, replyTo)))
+        // We don't need to catch exceptions thrown by await(): our code only completes it with valid results.
+        return withTimeoutOrNull(rpcTimeout) { replyTo.await() }
     }
 
-    suspend inline fun <reified T : ElectrumResponse> rpcCall(request: ElectrumRequest): T {
+    /** Send a request using timeouts, retrying once and giving up after that retry. */
+    private suspend inline fun <reified T : ElectrumResponse> rpcCall(request: ElectrumRequest): Either<ServerError, T> {
```

Comment:
We don't really have a choice, I think we have to retry to mitigate the case where we failed simply because we get disconnected and reconnected (as described in the comment of the `rpcCallWithTimeout` function).